### DISCUSSION TITLE
How to get the workflow to end the request and continue processing?

### DISCUSSION THREAD

**Discussion by weirdyang** [OrchardCMS/OrchardCore#16801] [Q&A]
I have a couple of workflows that take quite a bit of time to complete, and what I notice is that the request will not complete till the workflow has finished. Using a http response block does not work. the only way to end the request and continue running the workflow is to insert a signal event or custom event block.

However this means I will either need to trigger the signal via the url or trigger the custom event using a background task.

What are my alternatives?

**Comment by Piedone** [comment]
You can also add a Timer, and thus continue the rest of the workflow, after sending a response, in the background.

**Reply by weirdyang** [reply]
Yea that's what I'm sorta doing right now in my custom event block, but as i mentioned it still needs to be triggered

**Reply by Piedone** [reply]
I don't really understand what exactly you mean. The workflow needs to be triggered somehow to run in any case, what I understand you are using an HTTP request for. However, with a Timer task, there's no further trigger necessary, it'll run to completion in the background.

**Reply by weirdyang** [reply]
Yeap that what I meant, you still need to trigger the second event with a background task, so it needs to wait for the next interval before it can resume (eg which can be one second to one minute later), what I'm trying to achieve is the flow progressing without the need to wait for the second event to be triggered or even a second event

**Reply by Piedone** [reply]
Well, it either runs as part of the request, in the "foreground", but then it has to complete within the timeout of the request or otherwise risk being killed. The other option is to run it in the background, where it can run indefinitely (as long as the app itself runs). You can't have the first option so you need to use the second one.

**Comment by zl2fxy** [comment]
I think what you need is this method: HttpBackgroundJob.ExecuteAfterEndOfRequestAsync.
This method allows for running in the background without affecting the response, enabling an immediate response.

**Reply by zl2fxy** [reply]
If you want to use it in a workflow, you need to create a task to invoke it, or you can execute another workflow within this method.

**Reply by weirdyang** [reply]
I have considered this, but I will need to recreate the httprequest (e.g. the body,headers etc) or get the related info and use it as input for the workflow. Which isn't optimal as this will mean my existing workflows need to be redesigned. 

I ended up creating a background service with it's own queue to trigger the custom event instead, and letting the user know that any data from the http request will be lost after this activity and they should save all the info they need to a workflow property before this activity block. My custom activity will on execute queue the task to trigger the event, and return halt.

**Comment by SpikeXy** [comment]
Maybe HttpRequestTask need a TimeOut field like:

    using var cts = new CancellationTokenSource(TimeSpan.FromSeconds(TimeOut));

    var response = await httpClient.SendAsync(request, HttpCompletionOption.ResponseContentRead, cts.Token);
